Customisable frame metrics widget

MutantAura opened the pull request:
Follow-up to #2638 
It was mentioned that this feature may be nice to have and it was very basic in nature.

New setting in "General" which is an integer spinwheel for the number of decimal places the bottom render widget will display FPS to. This defaults to the value of 2 for new configuration files. Integer rounding value scales from 0-15 decimal places.
![image](https://user-images.githubusercontent.com/44103205/134261659-45199cd4-8ee6-476a-8d2c-9e54d16ee6c0.png)

See examples: 
2 decimal places
![image](https://user-images.githubusercontent.com/44103205/134261747-e826dec6-6a6c-481f-a51e-963056c1d3e3.png)
1 decimal place
![image](https://user-images.githubusercontent.com/44103205/134261774-f06580b1-5657-4389-a5bb-233df14bb669.png)
0 decimal places
![image](https://user-images.githubusercontent.com/44103205/134261810-82f6c25e-b5dc-4346-809d-a58a77824bf5.png)
15 decimal places (perhaps quite excessive...)
![image](https://user-images.githubusercontent.com/44103205/134261864-42212842-637f-42af-a671-0048dc2fce83.png)

There are some drawbacks. Using a rounding method to allow fully flexible control over this precision means if the final value is 0 then it will be truncated, cases where the user has specified 2DP and the value is 60.00 would be shortened to 60. Could possibly look into a `ToString()` method for each value but I've not thought of a way to do this efficiently yet as 15 if statements sounds messy... Might be more viable to do that if the values were restricted to 0-2 but I'm open to better ideas.
See 2DP defined but 0 has been truncated.
![image](https://user-images.githubusercontent.com/44103205/134262912-3cb71572-031d-49a2-802d-12a9152db0d0.png)

EliEron commented:
While I approve of this PR (as I prefer less precision personally) I agree with gdkchan that a dropdown would make far more sense. As you note yourself 15 digits of precision is quite ridiculous, honestly even 3 digits of precision seems overkill. So a short dropdown like the one gdkchan proposes seems far more intuitive to me.

MutantAura commented:
Updated implementation:

-Dropdown with common options instead of scroll wheel
-Fixed truncation of ending and leading 0's
-Fixes infinity frametime readings

MutantAura commented on Ryujinx.HLE/HLEConfiguration.cs:
```diff
@@ -109,6 +109,11 @@ public class HLEConfiguration
+        /// <summary>
+        /// Integer scale to adjust frame metric displayed value. Values from 0-4
+        /// </summary>
+        public static int fpsPrecision;
```
Didn't need any HLE references anyway so fixed by deletion.
Unsure why I added them in the first place.

MutantAura commented on Ryujinx.HLE/HLEConfiguration.cs:
```diff
@@ -163,7 +168,8 @@ public class HLEConfiguration
                                 bool                   ignoreMissingServices,
-                                AspectRatio            aspectRatio)
+                                AspectRatio            aspectRatio,
+                                int                    fpsPrecision)
```
As above, fixed by deletion.

MutantAura commented on Ryujinx.Headless.SDL2/Options.cs:
```diff
@@ -85,6 +85,9 @@ public class Options
         public int FsGlobalAccessLogMode { get; set; }
 
+        [Option("fps-precision", Required = false, Default = 0, HelpText = "Integer scale to adjust frame metric displayed value. Values from 0-4")]
+        public int fpsPrecision { get; set; }
```
Naming has been updated globally.

MutantAura commented on Ryujinx.Headless.SDL2/WindowBase.cs:
```diff
@@ -189,16 +191,62 @@ public void Render()
+                                break;
+                        }
+
+
```
Fixed

MutantAura commented on Ryujinx/Configuration/ConfigurationState.cs:
```diff
@@ -250,6 +257,7 @@ public SystemSection()
                 FsGlobalAccessLogMode         = new ReactiveObject<int>();
                 FsGlobalAccessLogMode.Event   += static (sender, e) => LogValueChange(sender, e, nameof(FsGlobalAccessLogMode));
                 AudioBackend                  = new ReactiveObject<AudioBackend>();
+                AudioBackend                  = new ReactiveObject<AudioBackend>();
```
hehe. fixed

MutantAura commented on Ryujinx/Ui/RendererWidgetBase.cs:
```diff
@@ -54,6 +54,10 @@ public abstract class RendererWidgetBase : DrawingArea
 
         private long _ticks = 0;
 
+        private string fpsString;
```
string naming updated

MutantAura commented on Ryujinx/Ui/RendererWidgetBase.cs:
```diff
@@ -417,17 +421,61 @@ public void Render()
                     {
                         string dockedMode = ConfigurationState.Instance.System.EnableDockedMode ? "Docked" : "Handheld";
-                        float scale = Graphics.Gpu.GraphicsConfig.ResScale;
+                        float scale       = Graphics.Gpu.GraphicsConfig.ResScale;
```
alignments should now be variable and value aligned

AcK77 commented on Ryujinx.Headless.SDL2/Options.cs:
```diff
@@ -85,6 +85,9 @@ public class Options
         public int FsGlobalAccessLogMode { get; set; }
 
+        [Option("fps-precision", Required = false, Default = 0, HelpText = "Integer scale to adjust frame metric displayed value. Values from 0-4")]
+        public static int FpsPrecision { get; set; }
```
It shouldn't be `static`

MutantAura replied:
I _think_ this is fixed...

AcK77 commented on Ryujinx.Headless.SDL2/OpenGL/OpenGLWindow.cs:
```diff
@@ -103,7 +103,7 @@ public void Dispose()
         private SDL2OpenGLContext _openGLContext;
 
-        public OpenGLWindow(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse) : base(inputManager, glLogLevel, aspectRatio, enableMouse)
+        public OpenGLWindow(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, int FpsPrecision) : base(inputManager, glLogLevel, aspectRatio, enableMouse, FpsPrecision)
```
```suggestion
        public OpenGLWindow(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, uint fpsPrecision) : base(inputManager, glLogLevel, aspectRatio, enableMouse, fpsPrecision)
```

AcK77 commented on Ryujinx.Headless.SDL2/Options.cs:
```diff
@@ -85,6 +85,9 @@ public class Options
         public int FsGlobalAccessLogMode { get; set; }
 
+        [Option("fps-precision", Required = false, Default = 0, HelpText = "Integer scale to adjust frame metric displayed value. Values from 0-4")]
+        public int FpsPrecision { get; set; }
```
```suggestion
        public uint FpsPrecision { get; set; }
```

The value is between 0 and 4 so a uint should be enough

AcK77 commented on Ryujinx.Headless.SDL2/Program.cs:
```diff
@@ -500,7 +500,7 @@ private static bool LoadApplication(Options options)
             Logger.RestartTime();
 
-            _window = new OpenGLWindow(_inputManager, options.LoggingGraphicsDebugLevel, options.AspectRatio, (bool)options.EnableMouse);
+            _window = new OpenGLWindow(_inputManager, options.LoggingGraphicsDebugLevel, options.AspectRatio, (bool)options.EnableMouse, (int)options.FpsPrecision);
```
```suggestion
            _window = new OpenGLWindow(_inputManager, options.LoggingGraphicsDebugLevel, options.AspectRatio, (bool)options.EnableMouse, options.FpsPrecision);
```

AcK77 commented on Ryujinx.Headless.SDL2/WindowBase.cs:
```diff
@@ -52,8 +54,9 @@ abstract class WindowBase : IHostUiHandler, IDisposable
 
         private AspectRatio _aspectRatio;
         private bool _enableMouse;
+        private int fpsPrecision;
```
```suggestion
        private uint _fpsPrecision;
```

AcK77 commented on Ryujinx.Headless.SDL2/WindowBase.cs:
```diff
@@ -52,8 +54,9 @@ abstract class WindowBase : IHostUiHandler, IDisposable
+        private int fpsPrecision;
 
-        public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse)
+        public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, int FpsPrecision)
```
```suggestion
        public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, AspectRatio aspectRatio, bool enableMouse, uint fpsPrecision)
```

AcK77 commented on Ryujinx.Headless.SDL2/WindowBase.cs:
```diff
@@ -67,6 +70,7 @@ public WindowBase(InputManager inputManager, GraphicsDebugLevel glLogLevel, Aspe
             _exitEvent = new ManualResetEvent(false);
             _aspectRatio = aspectRatio;
             _enableMouse = enableMouse;
+            fpsPrecision = FpsPrecision;
```
```suggestion
            _fpsPrecision = fpsPrecision;
```

AcK77 commented on Ryujinx.Headless.SDL2/WindowBase.cs:
```diff
@@ -188,17 +192,61 @@ public void Render()
                         string dockedMode = Device.System.State.DockedMode ? "Docked" : "Handheld";
-                        float scale = Graphics.Gpu.GraphicsConfig.ResScale;
+                        float  scale      = Graphics.Gpu.GraphicsConfig.ResScale;
+                        int    precision  = fpsPrecision;
```
You can remove this line

AcK77 commented on Ryujinx.Headless.SDL2/WindowBase.cs:
```diff
@@ -188,17 +192,61 @@ public void Render()
                             dockedMode += $" ({scale}x)";
                         }
 
+                        switch (precision)
```
```suggestion
                        switch (_fpsPrecision)
```

AcK77 commented on Ryujinx/Configuration/ConfigurationFileFormat.cs:
```diff
@@ -174,6 +174,11 @@ public class ConfigurationFileFormat
+        /// <summary>
+        /// Integer scale to adjust frame metric displayed value. Values from 0-4
+        /// </summary>
+        public int FpsPrecision { get; set; }
```
```suggestion
        public uint FpsPrecision { get; set; }
```

AcK77 commented on Ryujinx/Configuration/ConfigurationState.cs:
```diff
@@ -88,14 +88,21 @@ public ColumnSortSettings()
+            // <summary>
+            /// Integer scale to adjust frame metric displayed value. Values from 0-4
+            /// </summary>
+            public ReactiveObject<int> FpsPrecision { get; private set; }
```
```suggestion
            public ReactiveObject<uint> FpsPrecision { get; private set; }
```

AcK77 commented on Ryujinx/Configuration/ConfigurationState.cs:
```diff
@@ -88,14 +88,21 @@ public ColumnSortSettings()
+                EnableCustomTheme  = new ReactiveObject<bool>();
+                CustomThemePath    = new ReactiveObject<string>();
+                StartFullscreen    = new ReactiveObject<bool>();
+                FpsPrecision       = new ReactiveObject<int>();
```
```suggestion
                FpsPrecision       = new ReactiveObject<uint>();
```

AcK77 commented on Ryujinx/Ui/RendererWidgetBase.cs:
```diff
@@ -54,6 +54,10 @@ public abstract class RendererWidgetBase : DrawingArea
         private long _ticks = 0;
 
+        private string _fpsString;
+
```
Remove the extra empty line

AcK77 commented on Ryujinx/Ui/RendererWidgetBase.cs:
```diff
@@ -417,17 +421,61 @@ public void Render()
                         string dockedMode = ConfigurationState.Instance.System.EnableDockedMode ? "Docked" : "Handheld";
-                        float scale = Graphics.Gpu.GraphicsConfig.ResScale;
+                        float  scale      = Graphics.Gpu.GraphicsConfig.ResScale;
+                        int    precision  = ConfigurationState.Instance.Ui.FpsPrecision;
```
You can remove this line

AcK77 commented on Ryujinx/Ui/RendererWidgetBase.cs:
```diff
@@ -417,17 +421,61 @@ public void Render()
                             dockedMode += $" ({scale}x)";
                         }
 
+                        switch (precision)
```
```suggestion
                        switch (ConfigurationState.Instance.Ui.FpsPrecision)
```

AcK77 commented on Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -305,6 +306,7 @@ private SettingsWindow(MainWindow parent, Builder builder, VirtualFileSystem vir
             _resScaleCombo.SetActiveId(ConfigurationState.Instance.Graphics.ResScale.Value.ToString());
             _anisotropy.SetActiveId(ConfigurationState.Instance.Graphics.MaxAnisotropy.Value.ToString());
             _aspectRatio.SetActiveId(((int)ConfigurationState.Instance.Graphics.AspectRatio.Value).ToString());
+            _fpsPrecisionCombo.SetActiveId(ConfigurationState.Instance.Ui.FpsPrecision.Value.ToString());
```
```suggestion
            _fpsPrecisionCombo.SetActiveId(((int)ConfigurationState.Instance.Ui.FpsPrecision.Value).ToString());
```

AcK77 commented on Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -493,6 +495,7 @@ private void SaveSettings()
             ConfigurationState.Instance.Graphics.ShadersDumpPath.Value         = _graphicsShadersDumpPath.Buffer.Text;
             ConfigurationState.Instance.Ui.GameDirs.Value                      = gameDirs;
             ConfigurationState.Instance.System.FsGlobalAccessLogMode.Value     = (int)_fsLogSpinAdjustment.Value;
+            ConfigurationState.Instance.Ui.FpsPrecision.Value                  = int.Parse(_fpsPrecisionCombo.ActiveId);
```
```suggestion
            ConfigurationState.Instance.Ui.FpsPrecision.Value                  = uint.Parse(_fpsPrecisionCombo.ActiveId);
```

AcK77 reviewed (approved):
Lgtm! Thanks!
